File: procedural/queens.swift
#!/usr/bin/env swift

var size = 0
if C_ARGC > 1 {
    size = String.fromCString(C_ARGV[1])!.toInt()!
}

var row = [Bool](count: size, repeatedValue: true)
var up = [Bool](count: 2 * size - 1, repeatedValue: true)
var down = [Bool](count: 2 * size - 1, repeatedValue: true)
var board = [Int](count: size, repeatedValue: 0)

func queen(n :Int) {
    var i: Int

    if n >= size {
        for (i  = 0; i < size - 1; ++i) {
            print(board[i]); print(" ")
        }
        println(board[i])
    } else {
        for i in 0..<size {
            if (row[i] && up[n+i] && down[n-i+size-1]) {
                (row[i], up[n+i], down[n-i+size-1]) = (false, false, false);
                board[n] = i + 1;
                queen(n + 1);
                (row[i], up[n+i], down[n-i+size-1]) = (true, true, true);
            }
        }
    }
}

queen(0)

File: procedural/queens.rs
use std::env;

fn main() {
    let size: usize = env::args().collect::<Vec<String>>()[1].parse().unwrap();

    let mut row = vec![true; size];
    let mut up = vec![true; 2*size-1];
    let mut down = vec![true; 2*size-1];
    let mut board: Vec<i32> = vec![0; size];

    queen(0, size, &mut board, &mut row, &mut up, &mut down);
}

fn queen(n: usize, size: usize, board: &mut Vec<i32>, row: &mut Vec<bool>, up: &mut Vec<bool>, down: &mut Vec<bool>) {
    if n >= size {
        for i in 0..(size-1) {
            print!("{} ", board[i]);
        }
        println!("{}", board[size-1]);
    } else {
        for i in 0..size {
            if row[i] && up[n+i] && down[n+size-i-1] { 
                row[i] = false;
                up[n+i] = false;
                down[n+size-i-1] = false;
                board[n] = i as i32 + 1;
                queen(n+1, size, board, row, up, down);
                row[i] = true;
                up[n+i] = true;
                down[n+size-i-1] = true;
            }
        }
    }
}
